[ML] Add a model memory estimation endpoint for anomaly detection (#53507)
elastic/elasticsearch: docs/reference/ml/anomaly-detection/apis/estimate-model-memory.asciidoc

[role="xpack"]
[testenv="platinum"]
[[ml-estimate-model-memory]]

> === Estimate {anomaly-jobs} Model Memory API

Reviewer: ```suggestion
=== Estimate {anomaly-jobs} model memory API
```

++++
<titleabbrev>Estimate Model Memory</titleabbrev>
++++

> Estimates the model memory an analysis config is likely to need given

Reviewer: Trying to work the concept of jobs into this intro:
```suggestion
Estimates the model memory an {anomaly-job} is likely to need based on analysis configuration details and
```

cardinality estimates for the fields it references.

[[ml-estimate-model-memory-request]]
==== {api-request-title}

`POST _ml/anomaly_detectors/_estimate_model_memory`

[[ml-estimate-model-memory-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have `manage_ml` or
`manage` cluster privileges to use this API. See
<<security-privileges>>.

[[ml-estimate-model-memory-desc]]
==== {api-description-title}

> This API enables you to estimate the model memory and {anomaly-job}

Reviewer: > This API enables you....

I think this sentence repeats what we've already said in the intro so unless we have more info to add we can omit this section.

configuration will require before you create the job.

[[ml-estimate-model-memory-request-body]]
==== {api-request-body-title}

> For a list of the properties that you can specify in the `analysis_config`

Reviewer: ```suggestion
`analysis_config`::  (Required, object) For a list of the properties that you can specify in the `analysis_config`
```

component of the body of this API, see <<put-analysisconfig>>.

> 

Reviewer: We'll need descriptions of the other required objects too:
```suggestion

`max_bucket_cardinality`:: (Required, object) TBD

`overall_cardinality`:: (Required, object) TBD
```

[role="xpack"]
[testenv="platinum"]
[[ml-estimate-model-memory]]
=== Estimate {anomaly-jobs} model memory API
++++

> <titleabbrev>Estimate Model Memory</titleabbrev>

Reviewer: ```suggestion
<titleabbrev>Estimate model memory</titleabbrev>
```

++++

Estimates the model memory an {anomaly-job} is likely to need based on analysis
configuration details and cardinality estimates for the fields it references.

[[ml-estimate-model-memory-request]]
==== {api-request-title}

`POST _ml/anomaly_detectors/_estimate_model_memory`

[[ml-estimate-model-memory-prereqs]]
==== {api-prereq-title}

* If the {es} {security-features} are enabled, you must have `manage_ml` or
`manage` cluster privileges to use this API. See
<<security-privileges>>.

[[ml-estimate-model-memory-request-body]]
==== {api-request-body-title}

`analysis_config`::
(Required, object) For a list of the properties that you can specify in the
`analysis_config` component of the body of this API, see <<put-analysisconfig>>.

`max_bucket_cardinality`::
(Optional, object) Estimates of the highest cardinality in a single bucket

> that will be observed for influencer fields over the time period the job

Reviewer: Rewriting to avoid ending sentence in "for".  Alternatively "over the time period for which the job..."
```suggestion
that will be observed for influencer fields over the time period that the job
```

> analyzes data for. To produce a good answer values must be provided for

Reviewer: ```suggestion
analyzes data. To produce a good answer, values must be provided for
```

all influencer fields. It does not matter if values are provided for fields
that are not listed as `influencers`. If there are no `influencers` then

> `max_bucket_cardinality` can be safely omitted from the request.

Reviewer: Is "safely" required here?  It implies a danger that seems unfounded to me
```suggestion
`max_bucket_cardinality` can be omitted from the request.
```

`overall_cardinality`::
(Optional, object) Estimates of the cardinality that will be observed for

> fields over the whole time period the job analyzes data for. To produce

Reviewer: Per previous suggestion, I'd recommend trying to avoid ending with "for". E.g. either using "that" or "for which" instead:
```suggestion
fields over the whole time period that the job analyzes data. To produce
```

> a good answer values must be provided for fields referenced in the

Reviewer: ```suggestion
a good answer, values must be provided for fields referenced in the
```

`by_field_name`, `over_field_name` and `partition_field_name` of any
detectors. It does not matter if values are provided for other fields.
If no detectors have a `by_field_name`, `over_field_name` or

> `partition_field_name` then `overall_cardinality` can be safely omitted

Reviewer: Ditto previous suggestion around "safely"
```suggestion
`partition_field_name` then `overall_cardinality` can be omitted
```
